src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs

```diff
@@ -0,0 +1,68 @@
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using NuGet.Shared;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    internal sealed class PackageSearchMetadataCacheObject
+    {
+        private readonly IPackageSearchMetadata _packageSearchMetadata;
+        private readonly IPackageMetadataProvider _packageMetadataProvider;
+
+        public PackageSearchMetadataCacheObject(IPackageSearchMetadata packageSearchMetadata, IPackageMetadataProvider packageMetadataProvider)
+        {
+            _packageSearchMetadata = packageSearchMetadata;
+            _packageMetadataProvider = packageMetadataProvider;
+            AllVersionsContextInfo = GetVersionInfoContextInfoAsync();
+            PackageDeprecationMetadataContextInfo = GetPackageDeprecationMetadataContextInfoAsync();
+            DetailedPackageSearchMetadataContextInfo = GetDetailedPackageSearchMetadataContextInfoAsync();
+        }
+
+        public ValueTask<IReadOnlyCollection<VersionInfoContextInfo>> AllVersionsContextInfo { get; }
+        public ValueTask<PackageDeprecationMetadataContextInfo?> PackageDeprecationMetadataContextInfo { get; }
+        public ValueTask<PackageSearchMetadataContextInfo> DetailedPackageSearchMetadataContextInfo { get; }
+
+        public static string GetCacheId(string packageId, bool includePrerelease, IReadOnlyCollection<PackageSourceContextInfo> packageSources)
+        {
+            var hashCodeCombiner = new HashCodeCombiner();
+            hashCodeCombiner.AddSequence(packageSources);
+            hashCodeCombiner.AddStringIgnoreCase(packageId);
+            hashCodeCombiner.AddObject(includePrerelease.GetHashCode());
+            return hashCodeCombiner.CombinedHash.ToString(CultureInfo.InvariantCulture);
+        }
+
+        private async ValueTask<IReadOnlyCollection<VersionInfoContextInfo>> GetVersionInfoContextInfoAsync()
+        {
+            IEnumerable<VersionInfo> versions = await _packageSearchMetadata.GetVersionsAsync();
+            IEnumerable<Task<VersionInfoContextInfo>> versionContextInfoTasks = versions.Select(async v => await VersionInfoContextInfo.CreateAsync(v));
+            return await Task.WhenAll(versionContextInfoTasks);
+        }
+
+        private async ValueTask<PackageDeprecationMetadataContextInfo?> GetPackageDeprecationMetadataContextInfoAsync()
+        {
+            PackageDeprecationMetadata? deprecationMetadata = await _packageSearchMetadata.GetDeprecationMetadataAsync();
+            if (deprecationMetadata == null)
+            {
+                return null;
+            }
+            return NuGet.VisualStudio.Internal.Contracts.PackageDeprecationMetadataContextInfo.Create(deprecationMetadata);
```

Comment:
Can simplify type name by removing namespace?

Reply:
It complains about the property with the same name, will leave as is.